### PR TITLE
Bump libgit2 to 1.4.0

### LINKED ISSUE #804: Bump libgit2 to 1.4.0
https://github.com/libgit2/libgit2/releases/tag/v1.4.0 was just released and contains https://github.com/libgit2/libgit2/pull/6203, which we need for unblocking fast fetches in Cargo: https://github.com/rust-lang/cargo/issues/10078.

I tried naïvely bumping the libgit2 submodule like in #751 but `cargo test` has some test failures and several segfaults, so this may be a high-effort upgrade.

### PR DESCRIPTION
**Pull request by dtolnay**
Closes #804.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -120,7 +120,7 @@ pub use crate::reference::{Reference, ReferenceNames, References};
 pub use crate::reflog::{Reflog, ReflogEntry, ReflogIter};
 pub use crate::refspec::Refspec;
 pub use crate::remote::{
-    FetchOptions, PushOptions, Refspecs, Remote, RemoteConnection, RemoteHead,
+    FetchOptions, PushOptions, Refspecs, Remote, RemoteConnection, RemoteHead, RemoteRedirect,
```
Comment (dtolnay) [inline src/lib.rs:123]: API question: `RemoteRedirect` or `Redirect`? The libgit2 enum is called `git_remote_redirect_t`.

**Review by alexcrichton — APPROVED** [review]
👍 Thanks!

**Comment by ehuss** [comment]
Would it be possible to bump the version probe [here](https://github.com/rust-lang/git2-rs/blob/509fe5832ea35f6a670333da9e98ba5571419318/libgit2-sys/build.rs#L17)?

**Comment by extrawurst** [comment]
@ehuss @alexcrichton yeah lets bump that before we mark a release please

**Comment by extrawurst** [comment]
we probably need a release checklist in the readme

**Pull request merged**